### PR TITLE
add docker compilation to travis

### PR DESCRIPTION
**Pull request by ebbit1q**
my attempt at simply adding the dockerfile then running compilation inside it

### CODE REVIEW

**Comment by tooomm** [comment]
We would need travis to deploy the .dmg file from the 18.04 container to github releases as well.

**Comment by ebbit1q** [comment]
I'm not sure what you mean, aren't dmg files for mac? Note that this direction only uses docker for the compilation, it literally runs .ci/travis-compile.sh on the local repo (not a copy but it has full access and will leave the local files as if compiled on 18.04).

**Comment by ebbit1q** [comment]
It would be preferable to have the docker commands that are in compile.sh right now in travis.yaml (right now compile.sh with DOCKER_BUILDTYPE="Debug" will run compile.sh itself with BUILDTYPE="Debug" inside the docker container) but conditionally. Not to mention the travis container no longer needs dependencies itself, it only needs to build the docker image.

**Comment by tooomm** [comment]
I meant .deb of course.. :)

**Comment by ebbit1q** [comment]
Well it does.

**Comment by ebbit1q** [comment]
I mean it does that for 16.04 doesn't it? This is no different. Where does it export it as .deb exactly? I can't find it.

File: .ci/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+        build-essential \
+        g++ \
+        cmake \
+        libprotobuf-dev \
+        protobuf-compiler \
+        qt5-default \
+        qttools5-dev \
+        qttools5-dev-tools \
+        qtmultimedia5-dev \
+        libqt5multimedia5-plugins \
```
Comment (Daenyth) [inline .ci/Dockerfile]: Can you alphabetize the lines? It helps keep it manageable as it changes over time.

File: .ci/travis-compile.sh [hunk 2]
```diff
@@ -1,5 +1,18 @@
 #!/bin/bash
 
+if [[ $DOCKER_BUILDTYPE ]]; then
```
Comment (Daenyth) [inline .ci/travis-compile.sh]: Should this be `[[ -n $DOCKER_BUILDTYPE ]]` ? It seems to contain a string below

Comment (ebbit1q) [reply .ci/travis-compile.sh]: -n is the default operation, it doesn't need to be specified in bash.

**Comment by ebbit1q** [comment]
What is the -DCMAKE_PREFIX_PATH used for, and why is it set to a non existing location on linux? (I'll shield my eyes from the fact that it runs echo in a subshell)

**Comment by ctrlaltca** [comment]
The CMAKE_PREFIX_PATH parameter is used to specify one or more paths that cmake should try first while searching for libraries and binaries. The `/opt/qt5*/lib/cmake/` folder is where some PPAs we used installed custom versions of Qt5; i guess it was the ubuntu 12.x time and the version bundled in the distro was unusable, so we used a more recent version from a PPA.

**Comment by ebbit1q** [comment]
So it is no longer doing anything? What does it do on mac?

**Comment by ebbit1q** [comment]
Maybe a simple -r test?

**Comment by ctrlaltca** [comment]
https://github.com/Cockatrice/Cockatrice/blob/master/.ci/travis-compile.sh#L11
This is only used on travis, and it's the official way to check if you're running on osx.
Checking for "Darwin" is what you would normally do on a Makefile that's supposed to be run on everyone's box.

**Comment by ebbit1q** [comment]
Well that's the thing, the script is now run on a docker container in this pr, making it no longer a travis only script (even though the container runs in travis it's just nicer)

**Comment by ctrlaltca** [comment]
> the script is now run on a docker container in this pr

Right, now i get your point :)
Go on as you prefer, either define new env variables in travis, change how the script is working or even create a new compile script only used in docker builds, whatever fits you better

**Comment by ebbit1q** [comment]
The idea is to have as much similarity between the docker build and the normal build as possible, making sure there's no mistakes there.

File: .ci/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,19 @@
+FROM ubuntu:bionic
+
+RUN apt-get update && apt-get install -y --no-install-recommends \
```
[the remark this answers is not in the document]
Comment (ebbit1q) [reply .ci/Dockerfile]: Do we actually care about this?

**Comment by ebbit1q** [comment]
separated clangify changes to separate pr: #3435

**Comment by ebbit1q** [comment]
Happy travis, happy pr.

File: .travis.yml [hunk 4]
```diff
@@ -4,50 +4,65 @@ cache: ccache
...
-    os: linux
-    dist: xenial
-    group: stable
-    env: BUILDTYPE=Debug
-  - name: Ubuntu (Release)
+    services: docker
+    before_install: docker build -t img .ci
+    script: docker run --mount "type=bind,source=$(pwd),target=/src" -w="/src" img
+          bash .ci/travis-compile.sh --server "Debug"
+  - name: Ubuntu Bionic (Debug)
```
Comment (ctrlaltca) [inline .travis.yml]: I guess this is a typo, should be `(Release)`

File: .travis.yml [hunk 5]
```diff
@@ -4,50 +4,65 @@ cache: ccache
...
     if: (branch = master AND NOT type = pull_request) OR tag IS present
+    services: docker
+    script: docker build -t img .ci &&
+        docker run --mount "type=bind,source=$(pwd),target=/src" -w="/src" img
+          bash .ci/travis-compile.sh --server "Release"
```
Comment (ctrlaltca) [inline .travis.yml]: Do we need `--package` here, so that it can be uploaded it to the github release?

File: .travis.yml [hunk 6]
```diff
@@ -4,50 +4,65 @@ cache: ccache
...
-
-script: bash ./.ci/travis-compile.sh
+    before_install:
+      - brew update
+      - brew update
+      - brew install ccache
+      - brew unlink python # protobuf python2 install requires this link to be removed
+      - brew install protobuf
+      - brew install qt
+    script: bash ./.ci/travis-compile.sh --server --install "Release"
```
Comment (ctrlaltca) [inline .travis.yml]: Do we need --package here, too, so that it can be uploaded it to the github release?

File: .travis.yml (hunk 4 above)
Comment (ebbit1q) [reply .travis.yml]: Yes it is, I had to copy all the configs around, this is why we hate WET!

**Comment by ebbit1q** [comment]
I've separated the changes to clangify.sh in a separate pr! Please look at #3435 first!

File: .ci/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
```
[the remark this answers is not in the document]
Comment (Daenyth) [reply .ci/Dockerfile]: ^ I disabled that lint now, I disagree for this file

File: .ci/travis-dependencies.sh [hunk 8]
```diff
@@ -1,9 +0,0 @@
-#!/bin/bash
-
-if [[ $TRAVIS_OS_NAME == "osx" ]] ; then
-  brew update
```
Comment (Daenyth) [inline .ci/travis-dependencies.sh]: Why remove the OSX brew instructions?

Comment (Daenyth) [reply .ci/travis-dependencies.sh]: I see, moved to travis config

Comment (tooomm) [reply .ci/travis-dependencies.sh]: I think the idea with the external script was to keep the config file itself slimmer and cleaner. Ctrl introduced it back then.

**Comment by ebbit1q** [comment]
Please review this pr.

**Comment by ebbit1q** [comment]
There is no cp step required, this approach gives the docker container full access to the local files and builds in exactly the same location.

**Comment by ctrlaltca** [comment]
> There is no cp step required, this approach gives the docker container full access to the local files and builds in exactly the same location.

Ohh, now i see the magic:
```
    script: docker run --mount "type=bind,source=$(pwd),target=/src" -w="/src" img
```

**Comment by ebbit1q** [comment]
This also means there is no need to clone the repo twice :)

**Comment by ctrlaltca** [comment]
@ebbit1q Make package on ubuntu failed: https://travis-ci.org/Cockatrice/Cockatrice/jobs/456005691#L3024
Looks like it's missing https://packages.ubuntu.com/bionic/file

**Comment by ctrlaltca** [comment]
Also, on mac you can `brew install protobuf --without-python@2`, avoid the `- brew unlink python` and speed up a bit the build

**Comment by ebbit1q** [comment]
Cool, do you want me to make a new pr?

**Comment by ctrlaltca** [comment]
go for it 👍

**Comment by ebbit1q** [comment]
fixed in #3441

**Pull request merged**